core/node/events/stream_reconciler.go

```diff
@@ -0,0 +1,459 @@
+package events
+
+import (
+	"context"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+
+	"github.com/towns-protocol/towns/core/contracts/river"
+	. "github.com/towns-protocol/towns/core/node/base"
+	. "github.com/towns-protocol/towns/core/node/protocol"
+	"github.com/towns-protocol/towns/core/node/storage"
+)
+
+// TODO: move constants from rpc/forwarder.go to shared package so they are available here.
```

Comment:
We can also add something like this to `protocol/helpers.go`

```go
func NoRequestForward[T any](r *connect.Request[T]) {
	r.Header().Set("X-River-No-Forward", "true")
}

func NewRequestWithOpts[T any](message *T, opts ...func(r *connect.Request[T])) *connect.Request[T] {
	req := connect.NewRequest(message)
	for _, f := range opts {
		f(req)
	}
	return req
}
```

And than create the request with:

```go
req := NewRequestWithOpts(&GetMiniblocksRequest{
		StreamId:      streamId[:],
		FromInclusive: fromInclusive,
		ToExclusive:   toExclusive,
	}, NoRequestForward)
```